addons/mail/models/mail_thread.py

```diff
@@ -1124,6 +1125,97 @@ def message_process(self, model, message, custom_values=None,
         thread_id = self._message_route_process(message, msg_dict, routes)
         return thread_id
 
+    def _get_safe_create_data(self, vals_list, is_remove_missing_ref=False):
+        """ Return cleaned create vals_list to avoid missing reference, creation of new record, deletion and update by:
+
+        - Removing set to any non-existing id for relations many2one and many2many if is_remove_missing_ref
+        - Removing set to non-existing fields
+        - Removing set to one2many relation (which will update the other side of the relation)
+        - Removing set that create sub-record
+        - Removing any command != LINK or SET (i.e. skipping CREATE, UPDATE, DELETE, UNLINK, CLEAR)
+
+        :param list|dict vals_list: value to be cleaned
+        :param bool is_remove_missing_ref: whether to remove missing reference or not (see above). If set to False, no
+        DB query is done and the method is fast.
+        """
+        if isinstance(vals_list, dict):
+            vals_list = [vals_list]
+
+        def filter_existing_ids(f_name, ids):
+            """ Filter existing ids if is_remove_missing_ref, otherwise return the ids as is.
+
+             :param str f_name: field name of this record
+             :param list ids: id of the target of the field relation
+             :return: If is_remove_missing_ref, returns only existing ids of the relation defined by f_name
+             (being active or not) otherwise returns the ids as is.
+             Note: Null (False, None, ...) are considered as existing ids.
+             """
+            if not is_remove_missing_ref:
+                return ids
+            ids_not_null = {c_id for c_id in ids if c_id}
+            if ids_not_null:
+                self.env.cr.execute(
+                    query=f"SELECT id FROM {self.env[self._fields[f_name].comodel_name]._table} WHERE id IN %(ids)s",
+                    params=dict(ids=tuple(ids_not_null)))
+                filtered_existing_ids_set = set(map(lambda r: r['id'], self.env.cr.dictfetchall()))
+            else:
+                filtered_existing_ids_set = set()
+            return [c_id for c_id in ids if not c_id or c_id in filtered_existing_ids_set]
+
+        new_vals_list = []
+        for vals in vals_list:
+            vals_filtered = dict()
+            for field_name, value in vals.items():
+                new_value = None
+                field = self._fields.get(field_name)
+                if not field or field.type == 'one2many':
```

Comment:
Seems heuristics could simplified maybe ? I think we support only many2one, one2many and many2many for validation (other should be ok), so can't we somehow call the convert to cache of thoe 3 field types ?

Reply:
(See comment above about convert_to_cache)